scripts/ci/build_emscripten.sh

```diff
@@ -34,55 +34,101 @@
-
-echo "Emscripten output size: $OUTPUT_SIZE"
+function build() {
+    local build_dir=$1
+    local pre_release=$2
+
+    cd $WORKSPACE
+
+    # shellcheck disable=SC2166
+    if [[ "$CIRCLE_BRANCH" = release || -n "$CIRCLE_TAG" || -n "$FORCE_RELEASE" || "$(git tag --points-at HEAD 2>/dev/null)" == v* ]]
+    then
+        echo -n >prerelease.txt
+    else
+        # Use the given prerelease string if provided
+        if [[ "$pre_release" =~ ^[a-zA-Z]+\.[0-9]{4}\.[0-9]{1,2}\.[0-9]{1,2}$ ]]; then
+            echo "$pre_release" >prerelease.txt
+        else
+            # Use last commit date rather than build date to avoid ending up with builds for
+            # different platforms having different version strings (and therefore producing different bytecode)
+            # if the CI is triggered just before midnight.
+            TZ=UTC git show --quiet --date="format-local:%Y.%-m.%-d" --format="ci.%cd" >prerelease.txt
+        fi
+    fi
+    if [ -n "$CIRCLE_SHA1" ]
+    then
+        echo -n "$CIRCLE_SHA1" >commit_hash.txt
+    fi
+
+    # Disable warnings for unqualified `move()` calls, introduced and enabled by
+    # default in clang-16 which is what the emscripten docker image uses.
+    # Additionally, disable the warning for unknown warnings here, as this script is
+    # also used with earlier clang versions.
+    # TODO: This can be removed if and when all usages of `move()` in our codebase use the `std::` qualifier.
+    CMAKE_CXX_FLAGS="-Wno-unqualified-std-cast-call"
+
+
+    mkdir -p "$build_dir"
+    cd "$build_dir"
+    emcmake cmake \
+        -DCMAKE_BUILD_TYPE=Release \
+        -DBoost_USE_STATIC_LIBS=1 \
+        -DBoost_USE_STATIC_RUNTIME=1 \
+        -DCMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS}" \
+        -DTESTS=0 \
+    ..
+    make soljson
+
+    cd ..
+    mkdir -p upload
+    scripts/ci/pack_soljson.sh "$build_dir/libsolc/soljson.js" "$build_dir/libsolc/soljson.wasm" upload/soljson.js
+    cp upload/soljson.js ./
+
+    OUTPUT_SIZE=$(ls -la soljson.js)
+
+    echo "Emscripten output size: $OUTPUT_SIZE"
+}
+
+function show_help() {
+cat << EOF
+Usage: ${0##*/} [-h|--help] [--build-dir DIR] [--prerelease release_type.YYYY.MM.DD]
+Build Solidity emscripten binary
+    -h | --help  Display this help message
+    --build-dir  The emscripten build directory
+    --prerelease The prerelease string in the format: 'release_type.YYYY.MM.DD'
+EOF
+}
+
+function main() {
+    local build_dir="emscripten_build"
+    local pre_release=""
+
+    while :; do
+        case "$1" in
+            -h|--help) show_help; exit 0;;
+            --build-dir)
+                if [[ -n "$2" ]]; then
+                    build_dir="$2"
+                    shift
+                fi
+                shift
+                ;;
+            --prerelease)
+                if [[ -n "$2" ]]; then
+                    pre_release="$2"
+                    shift
+                fi
+                shift
+                ;;
+            *) break;;
```

Comment:
I think that currently something like `--prerelease ""` will make the script use the default value. Same if you do `--prereleaes` without giving it a value. This seems wrong. I think that the first should just use an empty string, the second should be an error (you can just omit the option instead and doing this is likely to be a mistake on user's part).

Same for `--build-dir`, though there being able to set an empty value is likely not useful so it could be an error as well.

Reply:
Yes, I guess there is no use for an empty `pre_release`, so the idea was to fall back to the default if it is none or empty. But if you think it is useful to accept empty `pre_release` I can change it to do what you suggest.

Reply:
I'm not sure if it will be useful in practice. Probably not - I'm fine also with not accepting it as long as it does not cause a fallback to default. If you explicitly use an option, the script should do what the option says. If it does not make sense it should fail rather than guess what you meant.

Reply:
Right. I changed the script to:
1) Fail on none/empty options (i.e., if you use an option, it must have a value)
2) Fail on invalid options while still accepting all options as optional.

Reply:
Looks good.